run_easymesh_cert: download from owncloud by default and some more features and fixes

tomereli opened the pull request:
Enhancements to download_ipk.sh and run_easymesh_cert.sh:
- Add download from owncloud support (default)
- Add option to skip prplmesh upgrade and make it optional
- Add option to run certification for a specific device

Needs to be merged after #1050

Commit:
run_easymesh_cert.sh: remove unused VERBOSE_OPT

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
run_easymesh_cert.sh: reorder getopt options

- Move long options only to be the last options in the usage print
- Fix shellcheck error with getopt usage

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
download_ipk.sh: download from owncloud support

Add support for downloading from owncloud public artifacts URL instead
of gitlab which as it happens is really not reliable.
Since the CI is now uploading artifacts to owncloud by default, also set
the default to use owncloud.
Add an option --remote which allows the user to choose which remote to
download the ipk from - currently supporting gitlab (deprecated) and
owncloud (preferred).

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
run_easymesh_cert.sh: skip prplmesh upgrade option

- remove branch option since by default we are now downloading from
owncloud which has only latest build for master branch
- move prplmesh upgrade to a separate function and make it optional and
for rax40 device only
- do not fail the script if upgrade fails

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
run_easymesh_cert.sh: add device option

Add option to run easymesh certification on a different device.
For prplmesh devices (netgear-rax40, turris omnia and glinet-1300) the
upload path is the target device name.
Other devices upload path is certified/$TARGET_DEVICE.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
run_easymesh_cert.sh: fix upload_to_owncloud.sh path

upload_to_owncloud.sh path was changed to ci/owncloud which was not
updated in this script, so do this now.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
ci: owncloud_definitions: add browse_url api

Add browse_url function to owncloud_definitions.sh which calculates the
public url for a given relative path from the root folder.

Since browse URL is using a public URL as base, add associative array of
the supported public URLs to the owncloud_definitions.sh.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
ci: owncloud_definitions: make script source'able

Preparative commit.
The owncloud_definitions.sh can be sourced by other scripts.
Due to this, it should not define global variables such as scriptdir and
rootdir which are defined in the sourcing script, since it will override
the paths.

Instead, take advantage on the fact we're already bash compliant and use
the ${BASH_SOURCE[0]} variable to inline-source functions.sh without
defining any global variables.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
ci: upload_to_owncloud: upload runtime and size

Add upload runtime and size to the prints in non verbose mode.
These will be useful to analyze the time it takes to upload to owncloud.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
ci: upload_to_owncloud: move upload to a function

Add upload function and move the uploading from main to it, to make
main() more readable.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
ci: owncloud_definitions: download_artifact api

Preparative commit.
Add download_artifact api which will be used to download latest prplmesh
when running certification.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
ci: run_easymesh_cert: add browse url

Now that owncloud_definitions.sh provides an API for getting a public
browse URL, print it after uploading logs.

While at it, fix shellcheck wrong source path when source'ing
functions.sh.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
ci: run_easymesh_cert: provide full log path to upload_to_owncloud.sh

The upload_to_owncloud.sh accepts two arguments:
$1 remote relative path from the user home folder
$2 local path to a directory to upload

For example, say we want to upload 'test' directory to
'Nightly/netgear-rax40/' folder in the remote.
The result of 'upload_to_owncloud.sh Nightly/netgear-rax40 test' will be
that the test folder will be uploaded to Nightly/netgear-rax40/test, but
all other files and folders under Nightly/netgear-rax40 will be deleted,
which is obviousely not what we want.

This is actually a bug in the move mechanism introduced in commit e3d5736b
and will be fixed in a separate commit, but for now - workaround the
issue by explicitly specifying the folder name in the remote path.

This has a side effect that the remote path will eventually be
Nightly/netgear-rax40/test/test instead of Nightly/netgear-rax40/test,
but its a small price to pay so we can live with it for now.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

rmelotte commented on ci/owncloud/owncloud_definitions.sh:
```diff
@@ -117,6 +113,8 @@ OWNCLOUD_PUBLIC_URLS["artifacts"]="https://ftp.essensium.com/owncloud/index.php/
-    "$@"
-fi
+[[ "$_" != "$0" ]] || {
+    if [ ! "$#" = 0 ] ; then
```
>In addition, do run the $@ commands only if the script is not sourced,
since this produced another warning when sourced by
run_easymesh_cert.sh.

This is what this specific line was supposed to do, so if it doesn't work, the new condition should replace this one.
What doesn't work/what warning do you get without the `[[ "$_" != "$0" ]] || {` part?

A downside of this last condition is that you can no longer call the script with anything in front of the script invocation (try `bash ci/owncloud/owncloud_definitions.sh copy a b` or even `HELLO=WORLD ./ci/owncloud/owncloud_definitions.sh copy a b`.

tomereli replied:
So apparently this is a buggy way to do this, since it doesn't work when sourcing the script from a different script, I changed to use ${BASH_SOURCE[0]} = $0 condition to check if the script is being sourced.

rmelotte commented on tools/download_ipk.sh:
```diff
@@ -1,34 +1,37 @@
+
+# shellcheck source=functions.sh
+. "$rootdir/tools/functions.sh"
+# shellcheck source=ci/owncloud/owncloud_definitions.sh
```
The path looks wrong.

rmelotte commented on tools/download_ipk.sh:
```diff
@@ -1,34 +1,37 @@
 
 usage() {
-    echo "usage: $(basename "$0") [-hbd]"
+    echo "usage: $(basename "$0") [-hbdrp]"
```
Only `h` and `v` are defined as short options.

rmelotte commented on ci/run_easymesh_cert.sh:
```diff
@@ -12,8 +12,10 @@
-# shellcheck source=functions.sh
+# shellcheck source=../tools/functions.sh
 . "$rootdir/tools/functions.sh"
+# shellcheck source=ci/owncloud/owncloud_definitions.sh
```
The path is wrong, it has to be relative to the script directory.

rmelotte replied:
You marked it as resolves but Github still shows it as `ci/owncloud/owncloud_definitions.sh` instead of `owncloud/owncloud_definitions.sh`.
Maybe you didn't push it yet?

rmelotte reviewed (approved):
Only small comments (except maybe for the `"$_" != "$0"` condition) so approving

Commit:
ci: run_easymesh_cert: remove extra echo

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout commented on ci/run_easymesh_cert.sh:
```diff
@@ -20,12 +20,12 @@ usage() {
     echo "      -s|--ssh - target device ssh name (defined in ~/.ssh/config). (default: $TARGET_DEVICE_SSH)"
     echo "      --owncloud-upload - whether or not to upload results to owncloud. (default: false)"
     echo "      --owncloud-path - relative path in the owncloud server to upload results to. (default: $OWNCLOUD_PATH)"
+    echo "      --skip-upgrade - skip prplmesh upgrade for netgear-rax40 device (default: false)"
```
```suggestion
    echo "      --skip-upgrade - skip prplmesh upgrade (default: false)"
```

arnout commented on ci/run_easymesh_cert.sh:
```diff
@@ -70,28 +87,19 @@ main() {
-    "$TOOLS_PATH"/download_ipk.sh --branch "$BRANCH" ${VERBOSE:+ -v} || {
-        err "Failed to download prplmesh.ipk, abort"
-        exit 1
+    [ "$DEVICE" = "netgear-rax40" ] && [ "$UPGRADE_PRPLMESH" = "true" ] && {
```
I don't like hardcoding netgear-rax40 here... Which device do you want to exclude here? If it is Ubuntu, the exclude that one specifically. If it is UGW/RDKB, hm, I don't know...

arnout commented on ci/run_easymesh_cert.sh:
```diff
@@ -90,9 +105,9 @@ main() {
+    info "Device: $TARGET_DEVICE"
 
-    [ "$DEVICE" = "netgear-rax40" ] && [ "$UPGRADE_PRPLMESH" = "true" ] && {
+    [ "$UPGRADE_PRPLMESH" = "true" ] && {
```
Ah, well, never mind then - but then you shouldn't have added that condition in the previous commit :-)

Don't bother changing it.

arnout commented on ci/owncloud/owncloud_definitions.sh:
```diff
@@ -1,11 +1,7 @@
-. "$rootdir/tools/functions.sh"
-
+# shellcheck source=../../tools/functions.sh
+. "$(dirname "${BASH_SOURCE[0]}")/../../tools/functions.sh"
```
Excellent!

arnout commented on ci/owncloud/owncloud_definitions.sh:
```diff
@@ -105,6 +105,33 @@ browse_url() {
+    # $2 base_url
+    local download_url
+
+    download_url="$(browse_url "$(dirname "$1")" "$2"/download)&files=$(basename "$1")"
```
I believe this should be
```suggestion
    download_url="$(browse_url "${1%%/*}" "$2"/download)&files=${1#*/}"
```
i.e. break off the first component of the path only, not the last component.

Commit:
ci: owncloud_definitions: fix script source check

${BASH_SOURCE[0]} (or, more simply, $BASH_SOURCE[1] )
contains the (potentially relative) path of the containing script
in all invocation scenarios, notably also when the script is
sourced, which is not true for $0, and causing the current check to fail
to identify that the script is being sourced.
Fix this by comparing ${BASH_SOURCE[0]} to $0.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
ci: upload_to_owncloud: use ${BASH_SOURCE[0]} for sourcing

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
ci: upload_to_owncloud: recursively create dirs

Add create_dirs helper function which recuresively creates directories
in the remote using owncloud_definitions.sh create_dir.
Use create_dirs to create the remote path the user requested, as well as
the temporary path we use for uploading.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
ci: upload_to_owncloud: add upload direct option

In some cases it is not needed to upload to a temporary folder before
moving to the destination path, so add an option to directly upload to
the destination folder (default false).

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
ci: run_easymesh_cert.sh: use --direct when uploading

When uploading a folder to owncloud, we are interested in the logs even
if they fail to upload in full. In addition, using the direct option
also avoids the issue with copy creating a directory inside directory.
For these reasons, use --direct option when calling
upload_to_owncloud.sh.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

arnout commented on ci/owncloud/owncloud_definitions.sh:
```diff
@@ -92,7 +92,29 @@ copy() {
+    # Calculate and echo the browse url for a relative path
+    #
+    # $1 relative path from the base_url
+    # $2 base_url
```
(nitpick) rather than base_url, IMO it would make more sense to pass in "certification" or "artifacts" and dereference OWNCLOUD_PUBLIC_URLS from within this function. In that case, I'd also pass it as the first argument, and the relative path second. It looks nicer that way IMO.

arnout commented on ci/owncloud/owncloud_definitions.sh:
```diff
@@ -105,6 +105,33 @@ browse_url() {
+    local base_url download_path
+
+    base_url="${OWNCLOUD_PUBLIC_URLS["artifacts"]}"
+    download_url="$(download_url "$1" "$base_url")"
```
So, with my earlier suggestion, this would become
```suggestion
    local download_path

    download_url="$(download_url "artifacts" "$1")"
```

arnout commented on tools/download_ipk.sh:
```diff
@@ -1,34 +1,37 @@
     echo "      -v|--verbose - add verbosity"
     echo "      --branch  - branch to use (default master)"
     echo "      --device  - device to use (default netgear-rax40)"
+    echo "      --remote  - remote to use - owncloud or gitlab (default owncloud)"
```
Instead of specifying remote, I'd simply infer it from the branch name. If branch is master, use owncloud, else use gitlab.

Note BTW that AFAIU, the artifacts are *always* uploaded to owncloud, even for branches. It's just that the "latest" only refers to master. If we need a different one, we'd need to find in which directory the latest for that branch was uploaded.

tomereli replied:
Still I'd like the flexibility of downloading from gitlab, so I'd prefer to still have a `--remote` option.

arnout commented on ci/owncloud/owncloud_definitions.sh:
```diff
@@ -140,6 +140,7 @@ OWNCLOUD_PUBLIC_URLS["artifacts"]="https://ftp.essensium.com/owncloud/index.php/
 
-if [ ! "$#" = 0 ] ; then
+# Check if script is being sourced
+if [ "${BASH_SOURCE[0]}" = "${0}" ]; then
```
Oh yes, this works as well!

arnout commented on ci/owncloud/upload_to_owncloud.sh:
```diff
@@ -28,8 +28,24 @@ usage() {
+    # Recursively create directory path in remote
+    # $1 remote path to create
+    local path=
+    for dir in $(echo "$1" | tr "/" "\\n"); do
```
Doesn't shellcheck complain about this?

I think a more typical pattern would be
```suggestion
    echo "$1" | tr / '\n' | while read -r -s dir; do
```

tomereli replied:
IMO `for` loop is easier to read, why is while preferred?

tomereli replied:
Actually @rmelotte 's suggestion to use `IFS` is the best IMO, so changed to that one.

arnout reviewed (approved):
These scripts are becoming waaay to complicated for shell :-)

TIL: when you write a script, do it in Python right away, even if you think it will be a simple script.

arnout commented:
I'm not sure why run-tests failed... It's not one of the randomly occuring issues...

rmelotte commented on ci/owncloud/upload_to_owncloud.sh:
```diff
@@ -28,8 +28,24 @@ usage() {
+            create_dir "$path/$dir"
+            path="$path/$dir"
+        fi
+    done
```
You can also loop over $1 directly if you set IFS to '/':
```suggestion
    local path IFS
    path=
    IFS='/'
    for dir in $1; do
        if [ -z "$path" ]; then
            create_dir "$dir"
            path="$dir"
        else
            create_dir "$path/$dir"
            path="$path/$dir"
        fi
    done
    unset IFS
```
(`unset IFS` is only there "just in case" the shell do not implement `local` i.e. it's not needed for bash)

tomereli replied:
Done

rmelotte commented on ci/owncloud/upload_to_owncloud.sh:
```diff
@@ -28,8 +28,24 @@ usage() {
 }
 
+create_dirs() {
+    # Recursively create directory path in remote
```
Nice!

rmelotte commented on ci/owncloud/upload_to_owncloud.sh:
```diff
@@ -17,6 +17,7 @@ usage() {
     echo "      -h|--help - display this help."
     echo "      -v|--verbose - enable verbosity"
     echo "      -u|--url - webDAV URL (default https://ftp.essensium.com/owncloud/remote.php/dav/files/USERNAME)"
+    echo "      --direct - direct upload without first uploading to a temporary directory (default false)"
```
I still don't really understand why you'd prefer to upload directly there, but ok.

Not ok though: the help message says the default is false, but you set it to true.

tomereli replied:
It is mainly to avoid the dir/dir issue.. and it is confusing and unnecessary to upload to a temp location when uploading logs - if the upload fails at some point, the logs we did succeed to upload are still interesting, so direct upload IMO is the correct approach in this case.

tomereli replied:
The default is empty, which means it is not true. I will add a default so it is clearer.

rmelotte commented on ci/run_easymesh_cert.sh:
```diff
@@ -121,16 +121,16 @@ main() {
+            REMOTE_PATH="certified/$TARGET_DEVICE"
         fi
         info "Uploading $LOG_FOLDER to $OWNCLOUD_PATH/$REMOTE_PATH"
-        "$scriptdir"/owncloud/upload_to_owncloud.sh "$OWNCLOUD_PATH/$REMOTE_PATH" "$LOG_FOLDER" || {
```
Ok here you explain why you want the "direct" option.
Note that even if you do partial uploads to a temporary directory, you can still move the incomplete folder at the end if you want.

IMO, if the default behavior doesn't correspond to what you expect (i.e. it uploads by replacing the remote path you specified, instead of uploading /inside/ the remote path) we should change it instead of adding the `direct` option as a workaround.

To me you can merge it as it is now though.

tomereli replied:
I didn't want to change it so not to cause issues with the CI uploads. I think doing it in steps, and later removing the direct option is better.

rmelotte reviewed (approved):
Nothing blocking :+1:
